### DISCUSSION TITLE
modern.config中的source.exclude使用后启动报错

### DISCUSSION THREAD

**Discussion by NieShiyi** [web-infra-dev/rsbuild#1900] [General]
我采用的Menorepo和约定式路由组织项目，我希望在打包的时候通过source.exclude这个参数不要把我的一个文件见打包进去。

使用方式：在apps的项目下的modern.config.js中引用如下
···
import { defineConfig, appTools } from '@modern-js/app-tools';
import path from 'path';

module.exports = {
  runtime: {
    router: true,
    state: true,
  },
  html: {
    disableHtmlFolder: true,
    title: 'bio',
  },
  source: {
    mainEntryName: 'index',
    exclude: [/src\/routes\/__home\/blog/],
    // exclude: [path.resolve(__dirname, 'src/routes/newcrown')],
  },
  output: {
    distPath: {
      html: '',
      js: '',
      css: '',
    },
  },
  tools: {
    devServer: {
      proxy: {
        '/api': {
          target: 'http://xxxx/',
          changeOrigin: true,
          pathRewrite: { '^/api': '' },
        },
      },
    },
  },
  server: {
    ssr: false,
    port: 8099,
  },
  plugins: [appTools()],
};

···

爆错误如下：
这个文件夹下面的页面报错
Module parse failed: Unexpected token (20:9)
File was processed with these loaders:
 * ../../node_modules/.pnpm/@pmmmwh+react-refresh-webpack-plugin@0.5.10_react-refresh@0.14.0_webpack@5.89.0/node_modules/@pmmmwh/react-refresh-webpack-plugin/loader/index.js
You may need an additional loader to handle the result of these loaders.‘

**Comment by chenjiahan** [comment]
source.exclude 的作用是「跳过编译」而不是「跳过打包」，如果需要跳过打包，可以使用 externals 配置

**Reply by NieShiyi** [reply]
多谢，但是externals可以对node_modules中的依赖跳过打包，对我自己定义的约定路由的文件夹如何编写跳过呢？

**Comment by 9aoy** [comment]
@NieShiyi  听起来你是希望在使用 modern.js 约定式路由的时候，排除掉 routes/ 下的某个页面，不生成该页面？

**Reply by NieShiyi** [reply]
是的，我是希望排除routes下面的某一个文件夹下面的所有页面

**Reply by 9aoy** [reply]
根据我的理解，好像只能把 `layout.[tj]sx` / `page.[tj]sx` 改成  `_layout.[tj]sx` / `_page.[tj]sx`。

@yimingjfe 看下有更好的做法么，支持通过配置排除么

![20240320-180053](https://github.com/web-infra-dev/rsbuild/assets/22373761/cfd52c97-9739-4cb7-b49e-a81702bfdc0b)

**Reply by yimingjfe** [reply]
文件不以 layout 和 page 命名文件不可以吗？

**Reply by NieShiyi** [reply]
按照他的约定式路由定义是要按照这样的组织方式，只是我觉得他竟然用这种方式生成路由，针对这种方式应该有自定义的打包配置，可以跳过生成这些路由，并且不打包对应的文件。只是我没有找到

**Reply by yimingjfe** [reply]
你这里什么场景需要配置某个文件夹跳过生成某些路由呢？以特殊文件(layout, page)约定命名也是为了支持 `routes` 下可以有非路由文件。

**Reply by NieShiyi** [reply]
以特殊文件(layout, page)约定命名也是为了支持 routes 下可以有非路由文件。你的意思是说不用page.js 用index这种其他的命名吗？可以一试。

**Reply by NieShiyi** [reply]
我是想要打包的时候通过ci配置，打包的时候走不同的配置文件，然后可以在配置文件中控制跳过。或者有其他方式跳过也可以。

**Reply by yimingjfe** [reply]
是的，可以使用 index 这种命名的，目前你的问题解决了吗
